File: python-fastapi/app/routers/tools.py
"""Tools CRUD endpoints."""
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select, and_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.database import get_db
from app.models import Tool, Category
from app.schemas.tool import ToolCreate, ToolUpdate, ToolResponse

router = APIRouter()


@router.get(
    "/tools",
    response_model=List[ToolResponse],
    status_code=status.HTTP_200_OK,
    summary="List All Tools",
    description="Get a list of all tools with optional filtering by category, status, and vendor",
)
async def list_tools(
    category_id: Optional[int] = Query(None, description="Filter by category ID"),
    status_filter: Optional[str] = Query(None, alias="status", description="Filter by status (active/inactive/trial)"),
    vendor: Optional[str] = Query(None, description="Filter by vendor name"),
    search: Optional[str] = Query(None, description="Search in tool name or description"),
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=500, description="Maximum number of records to return"),
    db: AsyncSession = Depends(get_db),
):
    """
    Retrieve a list of tools with optional filtering.
    
    Args:
        category_id: Filter by category ID
        status_filter: Filter by status (active/inactive/trial)
        vendor: Filter by vendor name
        search: Search in tool name or description
        skip: Number of records to skip for pagination
        limit: Maximum number of records to return
        db: Database session
        
    Returns:
        List[ToolResponse]: List of tools matching the filters
    """
    # Build query with eager loading of category
    query = select(Tool).options(selectinload(Tool.category))
    
    # Apply filters
    filters = []
    if category_id is not None:
        filters.append(Tool.category_id == category_id)
    if status_filter:
        filters.append(Tool.status == status_filter.lower())
    if vendor:
        filters.append(Tool.vendor.ilike(f"%{vendor}%"))
    if search:
        filters.append(
            (Tool.name.ilike(f"%{search}%")) | (Tool.description.ilike(f"%{search}%"))
        )
    
    if filters:
        query = query.where(and_(*filters))
    
    # Apply pagination
    query = query.offset(skip).limit(limit)
    
    # Execute query
    result = await db.execute(query)
    tools = result.scalars().all()
    
    # Transform tools to include category name
    return [
        ToolResponse(
            **{**tool.__dict__, "category": tool.category.name if tool.category else None}
        )
        for tool in tools
    ]


@router.get(
    "/tools/{tool_id}",
    response_model=ToolResponse,
    status_code=status.HTTP_200_OK,
    summary="Get Tool by ID",
    description="Retrieve a specific tool by its ID",
)
async def get_tool(
    tool_id: int,
    db: AsyncSession = Depends(get_db),
):
    """
    Retrieve a single tool by ID.
    
    Args:
        tool_id: The tool ID
        db: Database session
        
    Returns:
        ToolResponse: The requested tool
        
    Raises:
        HTTPException: 404 if tool not found
    """
    query = select(Tool).options(selectinload(Tool.category)).where(Tool.id == tool_id)
    result = await db.execute(query)
    tool = result.scalar_one_or_none()
    
    if not tool:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tool with ID {tool_id} not found",
        )
    
    return ToolResponse(
        **{**tool.__dict__, "category": tool.category.name if tool.category else None}
    )


@router.post(
    "/tools",
    response_model=ToolResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create New Tool",
    description="Create a new tool in the system",
)
async def create_tool(
    tool_data: ToolCreate,
    db: AsyncSession = Depends(get_db),
):
    """
    Create a new tool.
    
    Args:
        tool_data: Tool creation data
        db: Database session
        
    Returns:
        ToolResponse: The created tool
        
    Raises:
        HTTPException: 404 if category not found, 400 if validation fails
    """
    # Verify category exists
    if tool_data.category_id:
        category_query = select(Category).where(Category.id == tool_data.category_id)
        category_result = await db.execute(category_query)
        category = category_result.scalar_one_or_none()
        
        if not category:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Category with ID {tool_data.category_id} not found",
            )
    
    # Create tool
    new_tool = Tool(**tool_data.model_dump())
    db.add(new_tool)
    await db.commit()
    await db.refresh(new_tool)
    
    # Load category relationship
    await db.refresh(new_tool, ["category"])
    
    return ToolResponse(**{**new_tool.__dict__, "category": new_tool.category.name if new_tool.category else None})


@router.put(
    "/tools/{tool_id}",
    response_model=ToolResponse,
    status_code=status.HTTP_200_OK,
    summary="Update Tool",
    description="Update an existing tool's information",
)
async def update_tool(
    tool_id: int,
    tool_data: ToolUpdate,
    db: AsyncSession = Depends(get_db),
):
    """
    Update an existing tool.
    
    Args:
        tool_id: The tool ID to update
        tool_data: Tool update data (only provided fields will be updated)
        db: Database session
        
    Returns:
        ToolResponse: The updated tool
        
    Raises:
        HTTPException: 404 if tool or category not found
    """
    # Get existing tool
    query = select(Tool).where(Tool.id == tool_id)
    result = await db.execute(query)
    tool = result.scalar_one_or_none()
    
    if not tool:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tool with ID {tool_id} not found",
        )
    
    # Verify category exists if updating
    if tool_data.category_id is not None:
        category_query = select(Category).where(Category.id == tool_data.category_id)
        category_result = await db.execute(category_query)
        category = category_result.scalar_one_or_none()
        
        if not category:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Category with ID {tool_data.category_id} not found",
            )
    
    # Update tool with provided fields
    update_data = tool_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(tool, field, value)
    
    await db.commit()
    await db.refresh(tool)
    
    # Load category relationship
    await db.refresh(tool, ["category"])
    
    return ToolResponse(**{**tool.__dict__, "category": tool.category.name if tool.category else None})


@router.delete(
    "/tools/{tool_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete Tool",
    description="Delete a tool from the system",
)
async def delete_tool(
    tool_id: int,
    db: AsyncSession = Depends(get_db),
):
    """
    Delete a tool.
    
    Args:
        tool_id: The tool ID to delete
        db: Database session
        
    Raises:
        HTTPException: 404 if tool not found
    """
    # Get existing tool
    query = select(Tool).where(Tool.id == tool_id)
    result = await db.execute(query)
    tool = result.scalar_one_or_none()
    
    if not tool:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tool with ID {tool_id} not found",
        )
    
    await db.delete(tool)
    await db.commit()

File: src/routes/tools.js
/**
 * Tools CRUD routes
 */
import express from 'express';
import { pool } from '../database/connection.js';
import { asyncHandler, AppError } from '../middleware/errorHandler.js';

const router = express.Router();

// GET /api/tools - List tools with filters
router.get('/', asyncHandler(async (req, res) => {
  const {
    category_id,
    status,
    vendor,
    search,
    skip = 0,
    limit = 100,
  } = req.query;

  let query = `
    SELECT t.*, c.name as category
    FROM tools t
    LEFT JOIN categories c ON t.category_id = c.id
    WHERE 1=1
  `;
  
  const params = [];
  let paramCount = 1;

  if (category_id) {
    query += ` AND t.category_id = $${paramCount++}`;
    params.push(category_id);
  }

  if (status) {
    query += ` AND t.status = $${paramCount++}`;
    params.push(status);
  }

  if (vendor) {
    query += ` AND t.vendor ILIKE $${paramCount++}`;
    params.push(`%${vendor}%`);
  }

  if (search) {
    query += ` AND (t.name ILIKE $${paramCount++} OR t.description ILIKE $${paramCount})`;
    params.push(`%${search}%`, `%${search}%`);
    paramCount++;
  }

  query += ` ORDER BY t.created_at DESC LIMIT $${paramCount++} OFFSET $${paramCount}`;
  params.push(limit, skip);

  const result = await pool.query(query, params);
  res.json(result.rows);
}));

// GET /api/tools/:id - Get tool by ID
router.get('/:id', asyncHandler(async (req, res) => {
  const { id } = req.params;

  const result = await pool.query(
    `SELECT t.*, c.name as category
     FROM tools t
     LEFT JOIN categories c ON t.category_id = c.id
     WHERE t.id = $1`,
    [id]
  );

  if (result.rows.length === 0) {
    throw new AppError(`Tool with ID ${id} not found`, 404);
  }

  res.json(result.rows[0]);
}));

// POST /api/tools - Create new tool
router.post('/', asyncHandler(async (req, res) => {
  const {
    name,
    description,
    vendor,
    website_url,
    category_id,
    monthly_cost,
    owner_department,
    status = 'active',
  } = req.body;

  // Basic validation
  if (!name || !vendor || !category_id || monthly_cost === undefined || !owner_department) {
    throw new AppError('Missing required fields', 400);
  }

  const result = await pool.query(
    `INSERT INTO tools (name, description, vendor, website_url, category_id, monthly_cost, owner_department, status)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
     RETURNING *`,
    [name, description, vendor, website_url, category_id, monthly_cost, owner_department, status]
  );

  res.status(201).json(result.rows[0]);
}));

// PUT /api/tools/:id - Update tool
router.put('/:id', asyncHandler(async (req, res) => {
  const { id } = req.params;
  const updates = req.body;

  // Check if tool exists
  const existing = await pool.query('SELECT * FROM tools WHERE id = $1', [id]);
  if (existing.rows.length === 0) {
    throw new AppError(`Tool with ID ${id} not found`, 404);
  }

  const fields = [];
  const values = [];
  let paramCount = 1;

  Object.keys(updates).forEach((key) => {
    if (updates[key] !== undefined) {
      fields.push(`${key} = $${paramCount++}`);
      values.push(updates[key]);
    }
  });

  if (fields.length === 0) {
    throw new AppError('No fields to update', 400);
  }

  fields.push(`updated_at = CURRENT_TIMESTAMP`);
  values.push(id);

  const query = `UPDATE tools SET ${fields.join(', ')} WHERE id = $${paramCount} RETURNING *`;
  const result = await pool.query(query, values);

  res.json(result.rows[0]);
}));

// DELETE /api/tools/:id - Delete tool
router.delete('/:id', asyncHandler(async (req, res) => {
  const { id } = req.params;

  const result = await pool.query('DELETE FROM tools WHERE id = $1 RETURNING *', [id]);

  if (result.rows.length === 0) {
    throw new AppError(`Tool with ID ${id} not found`, 404);
  }

  res.status(204).send();
}));

export default router;
